[Infra UI] Add date picker to asset details

crespocarlos opened the pull request:
closes: https://github.com/elastic/kibana/issues/160378

## Summary

This PR adds a date-picker to the Asset Details (full page and flyout)

**Flyout**
<img width="1435" alt="image" src="https://github.com/elastic/kibana/assets/2767137/a1adb3b0-b3d5-4ca3-9076-6dafd44eebbc">


**Full page view**
<img width="1422" alt="image" src="https://github.com/elastic/kibana/assets/2767137/ad14764b-fdc2-4773-b8d0-7f2e271d89a5">

**Setting a date range**

https://github.com/elastic/kibana/assets/2767137/4c41d9ee-6c18-45af-9bda-3cae6f372b7b

**Brush event filter**

https://github.com/elastic/kibana/assets/2767137/5830dcd7-7d66-45ed-a463-15bb79607f2a

### Bonus

The uptime link was removed since the UI won't work.

### How to test

- Start a local Kibana instance
- Start metricbeat with system module enabled
- Navigate to `Infrastructure` > `Hosts`
  - Click on a host name and check if the new Node Details page opened
      - Check if the date range is set with the same date range from the Host Detail view
- Navigate to `Infrastructure` > `Hosts`
   -  Open the flyout
   -  Click on Open as page and check if the new Node Details page opened with the same date range from the Host Detail view
- Navigate to Infrastructure > Inventory
      - With Show: Host selected, click on a waffle item to open the flyout, click on Open as page and check if the new Node Details page opened with the same date range from the Inventory UI

Validate if the date picker value is applied to all tabs that have the component (Overview, Metadata, Processes, Logs, Anomalies)
Validate if the date range changes when selecting a time range in the charts (in the flyout, the date range from the Hosts View must not change)

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

crespocarlos commented on x-pack/plugins/infra/public/components/asset_details/tabs/processes/processes.tsx:
```diff
@@ -90,87 +94,106 @@ export const Processes = () => {
-          <EuiIconTip
-            aria-label={i18n.translate(
-              'xpack.infra.metrics.nodeDetails.processesHeader.tooltipLabel',
+      <EuiFlexGroup direction="column" gutterSize="m">
```
Changed to use flexbox instead of having `<EuiSpacer size="m" />` to separate each section

crespocarlos commented on x-pack/plugins/infra/public/components/asset_details/tabs/processes/processes_table.tsx:
```diff
@@ -218,14 +228,6 @@ const ProcessesTableBody = ({ items, currentTime }: TableBodyProps) => (
   </>
 );
 
-const StyledTableBody = euiStyled(EuiTableBody)`
```
euiStyled is marked as deprecated.

crespocarlos commented on x-pack/plugins/infra/public/components/asset_details/tabs/processes/processes_table.tsx:
```diff
@@ -302,10 +304,25 @@ const columns: Array<{
-  overflow: hidden;
-  text-overflow: ellipsis;
-`;
+const CodeLine = ({ command }: { command: string }) => {
```
euiStyled is deprecated. Created this new component in order to replace it.

jennypavlova replied:
Thanks for changing those!

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/hosts/components/host_details_flyout/flyout_wrapper.tsx:
```diff
@@ -21,14 +21,14 @@ export interface Props {
       asset={{ id: name, name }}
       assetType="host"
-      dateRange={searchCriteria.dateRange}
+      dateRange={parsedDateRange}
```
With this change, we make sure the flyout will always open with the same date range. Previously, if the date picker had relative dates, each time the flyout opened a new absolute date would be calculated.

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/inventory_view/components/ml/anomaly_detection/anomalies_table/anomalies_table.tsx:
```diff
@@ -159,40 +161,59 @@ const AnomalyActionMenu = ({
+        align-self: center;
+      `}
+    >
+      <EuiEmptyPrompt
```
Replacing with an EUI component

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/inventory_view/components/ml/anomaly_detection/anomalies_table/anomalies_table.tsx:
```diff
@@ -436,73 +455,77 @@ export const AnomaliesTable = (props: Props) => {
-          showUpdateButton={false}
-          onTimeChange={onTimeChange}
-          width="full"
+      {!hideDatePicker && (
+        <EuiFlexItem grow={false}>
+          <EuiSuperDatePicker
+            start={dateRange.from}
+            end={dateRange.to}
+            showUpdateButton={false}
+            onTimeChange={onTimeChange}
+            width="full"
+          />
+        </EuiFlexItem>
```
Didn't want to mess with other components that use the anomalies table. We might want to refactor this component in the future and split content and date picker into separate components

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/metric_detail/asset_detail_page.tsx:
```diff
@@ -80,6 +80,17 @@ export const AssetDetailPage = () => {
     [parsedTimeRange.from, parsedTimeRange.to]
   );
 
+  // Retrocompatibility
+  const handleTabStateChange = ({ dateRange: newDateRange }: TabState) => {
+    if (newDateRange) {
+      setTimeRange({
+        from: newDateRange.from,
+        to: newDateRange.to,
+        interval: parsedTimeRange.interval,
+      });
+    }
+  };
```
We can remove this once we start storing the asset details state in the URL.

elasticmachine commented:
Pinging @elastic/infra-monitoring-ui (Team:Infra Monitoring UI)

jennypavlova commented on x-pack/plugins/infra/public/components/asset_details/hooks/use_date_range.ts:
```diff
@@ -5,33 +5,38 @@
+  initialDateRange: TimeRange;
+}
+
+export function useDateRangeProvider({ initialDateRange }: UseAssetDetailsStateProps) {
```
When we have the date range persisted in the URL (the case of the flyout for example) should we first get the date range from the URL value (passed to the asset details component) and if we don't have it to fallback to default (the value selected in the host view / the 15m default)?

crespocarlos replied:
You're completely right. I totally forgot about storing the date range in the URL state.

jennypavlova commented on x-pack/test/functional/apps/infra/node_details.ts:
```diff
@@ -61,35 +53,77 @@ export default ({ getPageObjects, getService }: FtrProviderContext) => {
+        ]);
+      });
+
+      describe('#Date picker', () => {
```
Nice ❤️

jennypavlova reviewed (commented):
Good job! 👏 The new date picker works fine and the date value persists between the tabs. I have some questions as I think I found 2 issues that we can check: 

In case the user changes the time range inside the flyout it persists when he opens the details page but when the back button is pressed the flyout loads with the default time range (not the selected one when the user left the page to see the details page):

https://github.com/elastic/kibana/assets/14139027/77290c50-3d1e-4393-b4a1-e2ae40d4af78

When the user selects the date inside the flyout and refreshes the page/shares the URL the time range does not persist

https://github.com/elastic/kibana/assets/14139027/65f04791-d359-4c84-9347-816f1ca40c8a

I am not sure about that but sometimes when using "now" as "to" date it is converted to relative data and sometimes it stays as "now". Usually, when the "from" date is changed it changes to the relative date instead of "now" so what should be the expected behavior (should we use "now" or we should always convert to the relative date?): 


https://github.com/elastic/kibana/assets/14139027/4666e3e0-5268-4600-8849-cf36c45ae90d

https://github.com/elastic/kibana/assets/14139027/8e823731-b973-4867-9467-de87c9b78266

<img width="1909" alt="Screenshot 2023-08-24 at 12 45 41" src="https://github.com/elastic/kibana/assets/14139027/35fd8d66-bb23-4b3b-8ed7-737a5dc6296a">

jennypavlova commented on x-pack/plugins/infra/public/components/asset_details/content/content.tsx:
```diff
@@ -5,33 +5,58 @@
+            FlyoutTabIds.OVERVIEW,
+            FlyoutTabIds.LOGS,
+            FlyoutTabIds.METADATA,
+            FlyoutTabIds.PROCESSES,
```
Never mind I saw that we plan to add a tooltip to explain that in [this issue ](https://github.com/elastic/kibana/issues/164594)so I guess it should be fine.

crespocarlos commented on x-pack/plugins/infra/public/components/asset_details/tabs/overview/metrics/metrics_grid.tsx:
```diff
@@ -50,6 +53,21 @@ export const MetricsGrid = React.memo(
+      ({
+        range,
+        preventDefault,
+      }: Parameters<NonNullable<LensEmbeddableInput['onBrushEnd']>>[0]) => {
```
Good point.

mykolaharmash commented on x-pack/plugins/infra/public/components/asset_details/content/content.tsx:
```diff
@@ -5,33 +5,58 @@
+          visibleFor={[
+            FlyoutTabIds.OVERVIEW,
+            FlyoutTabIds.LOGS,
+            FlyoutTabIds.METADATA,
```
Got it. Thank you for the explanation!

crespocarlos commented:
@elasticmachine merge upstream

jennypavlova commented on x-pack/plugins/infra/public/pages/metrics/metric_detail/hooks/use_metrics_time.ts:
```diff
@@ -64,9 +64,11 @@ export const useMetricsTime = () => {
   );
 
-  const updateTimeRange = useCallback((range: MetricsTimeInput) => {
+  const updateTimeRange = useCallback((range: MetricsTimeInput, parseDate = true) => {
```
Nit: maybe we can always pass a parsed date to this function so here we can simplify the logic, wdyt?

crespocarlos replied:
I don't want to change this function so much. We won't use it after this https://github.com/elastic/kibana/issues/164147 is done

jennypavlova reviewed (commented):
Thanks for the changes! I tried the navigation again and now the full page is setting the last one hour based on the "to" date - not the whole interval set in the flyout:

https://github.com/elastic/kibana/assets/14139027/f9e54a81-7275-4b5f-8ffb-3a43773ae428

Is that intentional or it should be changed?

But going back to the flyout is fixed (the selected dates before navigating to the full page view are shown) the URL sharing and persisting after refresh as well 🎉

crespocarlos commented:
> Thanks for the changes! I tried the navigation again and now the full page is setting the last one hour based on the "to" date - not the whole interval set in the flyout:
> 
>  nav_to_full_page.mov 
> Is that intentional or it should be changed?

 @jennypavlova , thanks for thoroughly testing things! The link was doing what the inventory does: end date - 1h. I've fixed it and added a test to validate it.

jennypavlova reviewed (approved):
LGTM 🚀 Thank you for the fixes!
